compiler/src/iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUInterfaces.cpp

```diff
@@ -7,7 +7,161 @@
+#include "mlir/Dialect/Utils/IndexingUtils.h"
 #include "mlir/IR/DialectImplementation.h"
 
+#define DEBUG_TYPE "iree-gpu-interfaces"
+#define DBGS() (llvm::dbgs() << "[" DEBUG_TYPE "]: ")
+
 #include "iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUInterfaces.cpp.inc"
+
+namespace mlir::iree_compiler::IREE::GPU {
+
+using ::mlir::iree_compiler::IREE::Codegen::TileSwizzle;
+
+//===----------------------------------------------------------------------===//
+// DataTiledMMAInterfaceAttr
+//===----------------------------------------------------------------------===//
+
+/// Returns the swizzled tile distribution shape after applying the swizzle
+/// permutation.
+static SmallVector<int64_t>
+getSwizzledDistributionShape(const TileSwizzle &swizzle) {
+  SmallVector<int64_t> shape;
+  for (TileSwizzle::ExpandShapeDimVectorType e : swizzle.expandShape) {
+    for (TileSwizzle::Dim d : e) {
+      shape.push_back(d.distributionSize);
+    }
+  }
+  applyPermutationToVector(shape, swizzle.permutation);
+  return shape;
+}
+
+void DataTiledMMAInterfaceAttr::getUndistributedTileTypes(
+    SmallVectorImpl<VectorType> &result) {
+  SmallVector<Type> elementTypes;
+  getElementTypes(elementTypes);
+  for (auto [i, elementType] : llvm::enumerate(elementTypes)) {
+    TileSwizzle swizzle = getTileSwizzle(i);
+    SmallVector<int64_t> shape;
+    for (TileSwizzle::ExpandShapeDimVectorType group : swizzle.expandShape) {
+      for (TileSwizzle::Dim d : group) {
+        shape.push_back(d.size);
+      }
+    }
+    applyPermutationToVector(shape, swizzle.permutation);
+    result.push_back(VectorType::get(shape, elementType));
+  }
+}
+
+void DataTiledMMAInterfaceAttr::getDistributedTileTypes(
+    SmallVectorImpl<VectorType> &result) {
+  SmallVector<Type> elementTypes;
+  getElementTypes(elementTypes);
+  auto getShape = [=](unsigned operandIndex) {
+    return sliceSwizzledShape(
+        getTileSwizzle(operandIndex), [](TileSwizzle::Dim d) {
+          return d.kind != TileSwizzle::Dim::Kind::CrossThread;
+        });
+  };
+  for (auto [i, elementType] : llvm::enumerate(elementTypes)) {
+    result.push_back(VectorType::get(getShape(i), elementType));
+  }
+}
+
+std::optional<::mlir::SmallVector<int64_t, 2>>
+DataTiledMMAInterfaceAttr::getUndistributedTileDimExpansion(
+    int64_t operandIndex, int64_t logicalDim) {
+  return std::nullopt;
+}
+
+LogicalResult DataTiledMMAInterfaceAttr::populateOperandOffsetsSizesStrides(
+    OpBuilder &builder, Location loc, uint32_t operandIndex, Value threadId,
+    ArrayRef<int64_t> permutation, SmallVectorImpl<OpFoldResult> &offsets,
+    SmallVectorImpl<OpFoldResult> &sizes,
+    SmallVectorImpl<OpFoldResult> &strides) {
+  TileSwizzle swizzle = getTileSwizzle(operandIndex);
+
+  LLVM_DEBUG({
+    DBGS() << "DataTiledMMAInterfaceAttr::populateOperandOffsetsSizesStrides\n";
+    DBGS() << "    operand: " << operandIndex << "\n";
+    DBGS() << "    swizzle: " << swizzle << "\n";
+  });
```

Comment:
I think I added the debug message; I missed the update when switching to `LDBG`. It can be `LDBG()` and you don't need to declare `DBGS()`. The header file is `DebugLog.h` but not `Debug.h`, fyi.

(I was bothered by long logs (ie debug_type + file + line prefix) in `LDBG`, but it is still a good feature.)